### PR TITLE
Add notebook and assets for Distilling Step By Step

### PR DESCRIPTION
**Pull request by anihm136**
Add notebook implementing the distillation approach from the [Distilling Step-By-Step](https://arxiv.org/pdf/2305.02301.pdf) paper. This method showcases a novel training method that can be used even with large teacher models that do not expose their internal parameters/logits, such as those provided by commercial APIs

### CODE REVIEW

**Comment by holtskinner** [comment]
@anihm136 General comment, it would probably make sense to move those JSON files into a public Cloud Storage bucket instead of having them in the repository. Both to reduce space in the repo and to make it easier for users of Colab.

We have a GCS bucket `gs://github-repo` for this purpose. I'll ask @polong-lin if we can add these files.

File: language/tuning/distilling_step_by_step/Distillation.ipynb [hunk 1]
```diff
@@ -0,0 +1,606 @@
+{
```
Comment (holtskinner) [inline language/tuning/distilling_step_by_step/Distillation.ipynb:1]: <p>Maybe add an inline link to the PaLM site when you mention it?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/276/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (holtskinner) [inline language/tuning/distilling_step_by_step/Distillation.ipynb:1]: <p>Do you need to install anything for this?</p><p> </p><p>If so, add a section at the top for installation of libraries. You can see an example in this notebook.</p><p> </p><p><a href="https://github.com/GoogleCloudPlatform/generative-ai/blob/main/language/getting-started/intro_palm_api.ipynb" target="_blank">https://github.com/GoogleCloudPlatform/generative-ai/blob/main/language/getting-started/intro_palm_api.ipynb</a></p><p> </p><p>Also, be sure to follow the structure of the intro for that notebook. Add the links to Colab/Workbench, list any billable components used.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/276/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (holtskinner) [inline language/tuning/distilling_step_by_step/Distillation.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #4.    def tokenize_function(examples):
</pre><p>nit: Can you add parameter and return value types throughout?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/276/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

**Comment by holtskinner** [comment]
General Comment: I like what you have here, but the notebooks in this repository are intended to use Google Cloud components (specifically Vertex AI). Would it be possible to edit the structure to use models hosted on Vertex AI? That would also make this unique compared to the existing code in https://github.com/google-research/distilling-step-by-step

For example, T5-Flan is available in Model Garden.

https://cloud.google.com/vertex-ai/docs/start/explore-models#:~:text=Quickstart-,T5%2DFLAN,-Language

**Comment by anihm136** [comment]
Finally back from vacation, I have started making the suggested changes. WRT using Google Cloud resources - I don't think I can use the models from Model Garden here, since I need a model checkpoint and not a hosted model. I can, however, add a section at the end to deploy the distilled model onto Vertex AI. Would that do?

File: language/tuning/distilling_step_by_step/Distillation.ipynb (hunk 1 above)
Comment (anihm136) [reply language/tuning/distilling_step_by_step/Distillation.ipynb]: <p>With the new intro structure, I have added a link to PaLM2 in the Overview section</p>

Comment (anihm136) [reply language/tuning/distilling_step_by_step/Distillation.ipynb]: <p>Done</p>

Comment (anihm136) [reply language/tuning/distilling_step_by_step/Distillation.ipynb]: <p>Done. I have added in most places, but had to omit while overriding HuggingFace classes as transformers does not have complete typing yet</p>

**Comment by holtskinner** [comment]
That makes sense, I think it would be helpful!

**Comment by anihm136** [comment]
Hi @holtskinner I have added the deployment steps as well. Tested to work on Colab as well as Workbench. One last change needed - there are a few more helper files (for Vertex AI image build and deploy) that I would like to add to the Cloud Storage bucket for ease of access - could you please help with that? The files are accessible for now at `gs://distillation_aux_files`, and a recursive copy is needed as there is a subdirectory in there. So `gsutil -m cp -r gs://distillation_aux_files/* gs://github-repo/distillation/` should work (with the appropriate permissions)

**Comment by holtskinner** [comment]
@anihm136 Thanks for making those updates!

I have just two more requests:

- Change the filename of this notebook to `distillation_step_by_step.ipynb`
  - We will likely be creating another notebook for distilling Google Cloud models, and I'd like them to have distinct names.
- Since the auxiliary files are all code for deployment (Python, Dockerfile), I'd recommend those go in a folder in this repository.
  - Create a folder `language/tuning/distillation/distillation_step_by_step/` and put the notebook and all of the auxiliary files in that directory.
  
Thanks!

**Comment by anihm136** [comment]
Wrt auxillary files - the reason I moved them out of the repo is because, when someone uses the Colab or Workbench link to open the notebook, it does not bring along the other files in the folder. So I felt the easiest way to get those files would be to pull them in from GCS with the others. I saw some other notebooks which have auxillary files, and they basically create the files as part of the notebook using `%writefile` (which I'd really like to avoid if possible). Thoughts?

**Comment by holtskinner** [comment]
I understand GCS being an easy way to transmit these files, however they really should be version controlled/tested/linted like the rest of the files.

You can also add a cell like this to download the relevant files from the GitHub Repo directly

You can create another child folder just for the auxiliary deployment files.

```
!wget https://raw.githubusercontent.com/GoogleCloudPlatform/generative-ai/main/language/tuning/distillation/distillation_step_by_step/deployment/*
```

**Review by holtskinner — APPROVED** [review]

**Comment by holtskinner** [comment]
Thanks for the contribution!
